crates/rattler_networking/src/s3_middleware.rs

```diff
@@ -0,0 +1,341 @@
+//! Middleware to handle `s3://` URLs to pull artifacts from an S3 bucket
+use core::panic;
+
+use async_trait::async_trait;
+use aws_config::BehaviorVersion;
+use aws_sdk_s3::presigning::PresigningConfig;
+use reqwest::{Request, Response};
+use reqwest_middleware::{Middleware, Next, Result as MiddlewareResult};
+use tracing::debug;
+use url::Url;
+
+use crate::{Authentication, AuthenticationStorage};
+
+/// Configuration for the S3 middleware.
+#[derive(Clone, Debug)]
+pub enum S3Config {
+    /// Use the default AWS configuration.
+    FromAWS,
+    /// Use a custom configuration.
+    Custom {
+        /// The endpoint URL to use for the S3 client.
+        endpoint_url: Url,
+        /// The region to use for the S3 client.
+        region: String,
+        /// Whether to force path style for the S3 client.
+        force_path_style: bool,
+    },
+}
+
+/// S3 middleware to authenticate requests.
+#[derive(Clone, Debug)]
+pub struct S3 {
+    auth_storage: AuthenticationStorage,
+    config: S3Config,
+    expiration: std::time::Duration,
+}
+
+/// S3 middleware to authenticate requests.
+#[derive(Clone, Debug)]
+pub struct S3Middleware {
+    s3: S3,
```

Comment:
we created this extra abstraction instead of putting auth_storage, config, expiration in `S3Middleware` directly since we plan to re-use it in `rattler-build upload s3`